### PR TITLE
Here's something may help.

### PR DESCRIPTION
**Pull request by okcd00**
I like this repository very much, Thanks for your idea sharing!       
As shown below, there are some updates I made, hope it brings you something.

## Correct Subdirectory Judge

**Before** 
```bash
> ls
bi_lstm/
candidate_module_list/
final_linear_list/
operator_embedding/
position_encoding/
relation_module_list/
word_embedding/
> mv word_embedding word_embeddings
mv: cannot move 'word_embedding' to a subdirectory of itself, 'word_embeddings'.
> exit()
```

** After **
``` bash
> mv word_embedding word_embeddings
> ls
bi_lstm/
candidate_module_list/
final_linear_list/
operator_embedding/
position_encoding/
relation_module_list/
word_embeddings/
> exit()
WARNING: there are pending changes that have not been saved to disk. Discard (y/N)?
```

## Allow loading on a CPU-only machine
**Before**
``` bash
$ ./graftr ./model_save_2018-01-18-175659.pt
Traceback (most recent call last):
  File "./graftr", line 573, in <module>
    CheckpointShell(sys.argv[1]).cmdloop()
  File "./graftr", line 238, in __init__
    self._state = torch.load(checkpoint_path)
  File "D:\Anaconda\lib\site-packages\torch\serialization.py", line 386, in load
    return _load(f, map_location, pickle_module, **pickle_load_args)
  File "D:\Anaconda\lib\site-packages\torch\serialization.py", line 573, in _load
    result = unpickler.load()
  File "D:\Anaconda\lib\site-packages\torch\serialization.py", line 536, in persistent_load
    deserialized_objects[root_key] = restore_location(obj, location)
  File "D:\Anaconda\lib\site-packages\torch\serialization.py", line 119, in default_restore_location
    result = fn(storage, location)
  File "D:\Anaconda\lib\site-packages\torch\serialization.py", line 95, in _cuda_deserialize
    device = validate_cuda_device(location)
  File "D:\Anaconda\lib\site-packages\torch\serialization.py", line 79, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.
```

**After**
```bash
$ ./graftr ./model_save_2018-01-18-175659.pt
Checkpoint loaded from ./model_save_2018-01-18-175659.pt.
Type help or ? to list commands.

> ls
bi_lstm/
candidate_module_list/
final_linear_list/
operator_embedding/
position_encoding/
relation_module_list/
word_embedding/
```

##  Add new CP operator.
```bash
$ ./graftr ./model_save_2018-01-18-175659.pt
Checkpoint loaded from ./model_save_2018-01-18-175659.pt.
Type help or ? to list commands.

> ls
bi_lstm/
candidate_module_list/
final_linear_list/
operator_embedding/
position_encoding/
relation_module_list/
word_embedding/
> cp word_embedding word_embeddings
> ls
bi_lstm/
candidate_module_list/
final_linear_list/
operator_embedding/
position_encoding/
relation_module_list/
word_embedding/
word_embeddings/
> cd word_embeddings
> ls
weight
> cd ..
> cd word_embedding
> ls
weight
```

What's more, is it possible to add a function that operates on slices?     
For example, `cp word_embeddings[0] word_embeddings[107]`, 
that helps when we want to copy the embedings from one token to another in BERT.

### CODE REVIEW

File: graftr [hunk 1]
```diff
@@ -457,6 +458,51 @@ class CheckpointShell(Cmd):
...
+    if src_node.full_name == dest:
+      return
+    if (dest + '/').startswith(src_node.full_name + '/'):
+      print(f'cp: cannot copy \'{m.group(1)}\' to a subdirectory of itself, \'{m.group(2)}\'.')
+      return
+    if dest_node is not None:
+      if not dest_node.is_directory:
+        print(f'cp: cannot overwrite \'{m.group(2)}\' with \'{m.group(1)}\'.')
+        return
+      clone_node = deepcopy(src_node)
```
Comment (sharvil) [inline graftr]: This `deepcopy` and the one below will make a copy of the entire tree including tensors. Instead, maybe you can try something like:

```python
class _Node:
  [...]
  def clone(self, clone_parent):
    parent = self.parent
    self.parent = None
    clone = deepcopy(self)
    clone.parent = clone_parent
    self.parent = parent
    return clone
```

and then call `src_node.clone(dest_node)` instead.

**Comment by sharvil** [comment]
These are all excellent changes. Thank you for suggesting them and for providing a PR!

About loading on a CPU-only machine: having `graftr` automatically load all tensors on CPU may not be the right behavior all the time. Suppose I have a model that's saved with the correct device information (e.g. GPU[1..n], TPU). If I load the model on a machine where GPUs/TPUs aren't available and then `save`, all devices will be silently converted to CPU.

Instead, I think we should have a command-line option to specify whether `graftr` should remap tensors to CPU if the target device isn't available or if it should exit with an error (default is error). I can make that change once we've merged your PR.

File: graftr (hunk 1 above)
Comment (okcd00) [reply graftr]: Thank you for your advice, I have learned a lot!  
I'll PR later after this update.

**Comment by sharvil** [comment]
Thanks, these changes look great!

You asked about operating on slices – that's a good idea as well. I was thinking of using `exec` or `eval` so the expressions can be more flexible. I haven't thought the idea through, but it seems like there's a way to make it work right.

**Comment by sharvil** [comment]
@okcd00 I just submitted a change that supports operating on slices (and a lot more!). You can use standard Python expressions to operate on tensors and other values. In your example, it would look something like:

```python
word_embeddings[0] = word_embeddings[107]
```

You could also do something more complex, like adding a little bit of Gaussian noise to the copied embedding vector:

```python
word_embeddings[0] = (0.125 * torch.randn_like(word_embeddings[0])) + word_embeddings[107]
```

**Comment by okcd00** [comment]
That's cool ! It's much more useful for transfer learning tasks now.

**Pull request merged**